src/ralph/schemas/edx/browser.py

```diff
@@ -227,18 +219,94 @@ def validate_event_textbook_pdf(self, data, **kwargs):
             return
 
+        event = self.parse_json_check_keys(data, {"chapter", "page", "name"})
+        self.check_chapter_page_name(data, event)
+
+    @validates_schema
+    def validate_event_textbook_pdf_thumbnail_navigated(self, data, **kwargs):
+        """Check that the event is a parsable JSON object with 4 keys:
+         `name`: string, `page`: integer > 0, `chapter`: url
+         `thumbnail_title`: string """
+        if data["name"] != "textbook.pdf.thumbnail.navigated":
+            return
+
+        event = self.parse_json_check_keys(
+            data, {"chapter", "page", "name", "thumbnail_title"}
+        )
+        self.check_chapter_page_name(data, event)
+        if not isinstance(event["thumbnail_title"], str):
+            raise ValidationError("thumbnail_title should be a string")
+
+    @validates_schema
+    def validate_event_textbook_pdf_zoom_buttons_changed(self, data, **kwargs):
+        """Check that the event is a parsable JSON object with 4 keys:
+        `name`: string, `page`: integer > 0, `chapter`: url,
+        `direction`: string (one of [`in`, `out`] or [`up`, `down`])
+        """
+        if data["name"] not in [
+            "textbook.pdf.zoom.buttons.changed",
+            "textbook.pdf.page.scrolled",
+        ]:
+            return
+
+        event = self.parse_json_check_keys(
+            data, {"chapter", "page", "name", "direction"}
+        )
+        self.check_chapter_page_name(data, event)
+        valid_directions = (
+            ["in", "out"] if data["name"][-7:] == "changed" else ["up", "down"]
+        )
+        if event["direction"] not in valid_directions:
+            raise ValidationError(
+                f"event.direction should be one of {valid_directions}"
+            )
+
+    @validates_schema
+    def validate_event_textbook_pdf_zoom_menu_changed(self, data, **kwargs):
+        """Check that the event is a parsable JSON object with 4 keys:
+        `name`: string, `page`: integer > 0, `chapter`: url,
+        `amaunt`: string
+        """
+        if data["name"] != "textbook.pdf.zoom.menu.changed":
+            return
+
+        event = self.parse_json_check_keys(data, {"chapter", "page", "name", "amaunt"})
+        self.check_chapter_page_name(data, event)
+        if event["amaunt"] not in BROWSER_EVENT_VALID_AMAUNT:
+            raise ValidationError(
+                f"amaunt should be one of {BROWSER_EVENT_VALID_AMAUNT}"
+            )
+
+    @validates("session")
+    def validate_session(self, value):  # pylint: no-self-use
+        """"Check session field empty or MD5_HASH_LEN chars long"""
+        if value != "" and len(value) != MD5_HASH_LEN:
+            raise ValidationError(
+                f"Session should be empty or {MD5_HASH_LEN} chars long (MD5)"
+            )
+
+    @staticmethod
+    def parse_json_check_keys(data, required_keys):
+        """Parse JSON string and check that it contains the keys in the
+        `keys_array` and no other keys
```

Comment:
sure?

Reply:
Good catch, thanks - required_keys is an set